src/mps/mps.jl

```diff
@@ -507,6 +507,155 @@ function sample(m::MPS)
+
+# Examples
+```julia
+N = 30
+m = 4
+
+s = siteinds("S=1/2",N)
+psi = randomMPS(s,m)
+Czz = correlator(psi,"Sz","Sz")
+
+s = siteinds("Electron",N; conserve_qns=true)
+psi = randomMPS(s,n->isodd(n) ? "Up" : "Dn",m)
+Cuu = correlator(psi,"Cdagup","Cup";site_range=2:8)
+```
+"""
+function correlation_matrix(psi::MPS,
+                            Op1::AbstractString,
+                            Op2::AbstractString;
+                            kwargs...)
+  N = length(psi)
+
+  site_range::UnitRange{Int} = get(kwargs,:site_range,1:N)
+  start_site = first(site_range)
+  end_site = last(site_range)
+
+  psi = copy(psi)
+  orthogonalize!(psi,start_site)
+  s = siteinds(psi)
+  onsiteOp = "$Op1*$Op2"
+  fermionic2 = has_fermion_string(Op2,s[1])
+  if fermionic2
+    Op1 = "$Op1*F"
+  end
+
+  # Nb = size of block of correlation matrix
+  Nb = end_site-start_site+1
+
+  C = zeros(Nb,Nb)
+
+  if start_site == 1
+    L = ITensor(1.)
+  else
+    lind = commonind(psi[start_site],psi[start_site-1])
+    L = delta(lind,lind')
+  end
+
+  for i=start_site:end_site-1
+    ci = i-start_site+1
+
+    Li = L*psi[i]
+
+    # Get j == i diagonal correlations
+    rind = commonind(psi[i],psi[i+1])
+    C[ci,ci] = scalar( Li*op(onsiteOp,s,i)*prime(dag(psi[i]),not(rind)) )
+
+    # Get j > i correlations
+    Li = Li*op(Op1,s,i)*dag(prime(psi[i]))
+    for j=i+1:end_site
+      cj = j-start_site+1
+      lind = commonind(psi[j],Li)
+      Li *= psi[j]
+
+      val = Li*op(Op2,s,j)*dag(prime(prime(psi[j],"Site"),lind))
+      C[ci,cj] = scalar(val)
+      C[cj,ci] = C[ci,cj]
+
+      if fermionic2
+        Li *= op("F",s,j)*dag(prime(psi[j]))
+      else
+        Li *= dag(prime(psi[j],"Link"))
+      end
+    end
+    L *= psi[i]*dag(prime(psi[i],"Link"))
+  end
+
+  # Get last diagonal element of C
+  i = end_site
+  lind = commonind(psi[i],psi[i-1])
+  C[Nb,Nb] = scalar( L*psi[i]*op(onsiteOp,s,i)*prime(prime(dag(psi[i]),"Site"),lind) )
+
+  return C
+end
+
+
+"""
+    expect(psi::MPS,ops::AbstractString...;kwargs...)
+
+Given an MPS `psi` and an operator name, returns
+a vector of the expected value of the operator on 
+each site of the MPS. If multiple operator names are
+provided, returns a tuple of expectation value vectors.
+
+# Optional Keyword Arguments
+- `site_range = 1:length(psi)`: compute expected values only for sites in the given range
+
+# Examples
+
+```julia
+N = 10
+
+s = siteinds("S=1/2",N)
+psi = randomMPS(s,8)
+Z = expect(psi,"Sz";site_range=2:6)
+
+s = siteinds("Electron",N)
+psi = randomMPS(s,8)
+dens = expect(psi,"Ntot")
+updens,dndens = expect(psi,"Nup","Ndn")
+```
+"""
+function expect(psi::MPS,
+                ops::AbstractString...;
+                kwargs...)
+  N = length(psi)
+  site_range::UnitRange{Int} = get(kwargs,:site_range,1:N)
+
+  Ns = length(site_range)
+  offset = first(site_range)-1
+
+  psi = copy(psi)
+  Nops = length(ops)
+  s = siteinds(psi)
+  ex = ntuple(n->zeros(Ns),Nops)
```

Comment:
I just realized that this assumes the result will be real. I suppose we are limiting this to Hermitian operators (otherwise an expectation value isn't defined)? Also this will probably error in general for complex cases since the result won't be exactly real, i.e.:
```julia
julia> x = zeros(4)
4-element Vector{Float64}:
 0.0
 0.0
 0.0
 0.0

julia> x[1] = 3.0 + 0.0im
3.0 + 0.0im

julia> x[1] = 3.0 + eps()im
ERROR: InexactError: Float64(3.0 + 2.220446049250313e-16im)
Stacktrace:
 [1] Real
   @ ./complex.jl:37 [inlined]
 [2] convert
   @ ./number.jl:7 [inlined]
 [3] setindex!(A::Vector{Float64}, x::ComplexF64, i1::Int64)
   @ Base ./array.jl:839
 [4] top-level scope
   @ REPL[9]:1
```
For now we could take the real part and check that the imaginary part is small, and limit to Hermitian operators.

This reminds me that we should also start to be careful about preserving element types by using generic code patterns like `zeros(real(eltype(psi)), Ns)`, in case people are using things like `Float32`, `BigFloat` or `Tropical` numbers, but maybe that is a story for another day.

Reply:
Good catch! I will try it out with some complex random MPS to make sure the final version works.

Yes, I do think it's reasonable to limit to Hermitian operators, so we can just take the real part.

I'll do some of the generic element type code too, good suggestion, but we will probably have to really test it out in the future to be sure all our code is fully generic.